content/blog/2022-05-23-awk-llvm/index.md

```diff
@@ -0,0 +1,135 @@
+
+## Points of Divergence (BRAWN vs AWK)
+For convenience, we restrict ourselves to a (large) subset of `AWK` with a few changes for ease of implementation. We find these choices defensible because they reduce burden on the parser and runtime, which while interesting, are not the focus of our project. Here are the points of divergence:
+
+1. We restrict I/O functionality. We support only the `print` function (no `printf` or `sprintf`). We only support reading input from `stdin` and writing output to `stdout`. Hence, we do not support built-in variables like `FNR`, `CONVFMT`, `OFMT`, and `FILENAME`.
+
+4. We require that all statement blocks like `if`, `for` etc. be followed by statements enclosed in `{}`.
+
+6. `AWK` uses the token `/` to indicate both `div` and the beginning/end of a regular expression. `BRAWN` requires that regular expressions begin with `/#` and end with `#/`.
+
+7. `AWK` offers a `bool`-returning membership query of the form `bird in birds`. `BRAWN` requires a different style: `[bird] in birds`.
+
+8. `AWK` offers string concatenation via the syntax `"concat" "these" "strings"`. `BRAWN` requires `"concat" @ "these" @ "strings"`.
+
+9. `AWK` allows functions to be declared anywhere in the program so long as they are outside of blocks. They can be called before they are declared. `BRAWN` requires that all functions be defined before any blocks are.
+
+
+# Evaluation
+
+## Correctness
+For a small selection of `AWK` programs that we can support in `BRAWN`, we compare our outputs to that of the standard `AWK` implementation. We select a series of `AWK` programs, generally a little more complex than simple one-line commands, from benchmarks previously used by
+* the `AWK`-like language [frawk](https://github.com/ezrosent/frawk)
+* a [user guide](https://www.math.utah.edu/docs/info/gawk_toc.html) for GNU `AWK`
+and convert them by hand into their equivalent `BRAWN` programs. The changes required are minor and are as described above. 
+
+The programs perform the following operations:
+
+| Name | Description |
+|------|-------------|
+| `math` [^6] | Compute means after grouping by the value of a column |
+| `dupword` [^5] | Detect consecutive uses of words in a text |
+| `word-count` [^5] | Report on the frequency of each word in a text |
+
+We use `turnt` to generate the outputs of `AWK`, and compare these with `BRAWN` outputs by eye. Our compiled programs pass these tests.
+
+## Benchmark
+We provide a brief benchmark against the standard GNU `AWK` implementation. Because startup times are high, only longer-running programs offer interesting data. In our case, that is `math.awk` versus `math.brawn`. We run our benchmarks on a MacBook Pro running MacOS 12; the `AWK` version is 5.1.1.
```

Comment:
I'm curious what the startup time is so high. Is it including the compilation time? Compiled C (or LLVM) programs usually start up pretty fast… if you are including the compilation time, I think it would be perfectly fair to *not* do so.